File: src/palate/agent/events.py
"""The typed events a run emits. The UI never sees raw provider JSON."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

from palate.providers.base import JSONObject


@dataclass(frozen=True, slots=True)
class RunStarted:
    type: Literal["run.started"] = "run.started"
    run_id: str = ""
    session_id: str = ""
    model: str = ""
    budget: JSONObject = field(default_factory=dict)


@dataclass(frozen=True, slots=True)
class TurnStarted:
    type: Literal["turn.started"] = "turn.started"
    turn: int = 0
    tools_offered: tuple[str, ...] = ()


@dataclass(frozen=True, slots=True)
class TextDelta:
    type: Literal["text.delta"] = "text.delta"
    text: str = ""
    channel: Literal["thinking", "answer"] = "answer"


@dataclass(frozen=True, slots=True)
class ToolCallProposed:
    type: Literal["tool.proposed"] = "tool.proposed"
    call_id: str = ""
    name: str = ""
    arguments: JSONObject = field(default_factory=dict)
    source: Literal["field", "content"] = "field"


@dataclass(frozen=True, slots=True)
class ToolCallStarted:
    type: Literal["tool.started"] = "tool.started"
    call_id: str = ""
    name: str = ""


@dataclass(frozen=True, slots=True)
class ToolCallFinished:
    type: Literal["tool.finished"] = "tool.finished"
    call_id: str = ""
    name: str = ""
    ok: bool = True
    # One line for the UI card, never the payload.
    summary: str = ""
    meta: JSONObject = field(default_factory=dict)
    latency_ms: int = 0
    undo_token: str | None = None


@dataclass(frozen=True, slots=True)
class PreferenceRecorded:
    type: Literal["preference.recorded"] = "preference.recorded"
    pref_id: int = 0
    label: str = ""
    polarity: str = ""
    hardness: str = ""
    affected_films: int = 0
    undo_token: str = ""


@dataclass(frozen=True, slots=True)
class GroundingChecked:
    type: Literal["grounding.checked"] = "grounding.checked"
    grounded_ratio: float = 1.0
    nli_available: bool = False
    unsupported: tuple[str, ...] = ()
    dropped_film_ids: tuple[int, ...] = ()


@dataclass(frozen=True, slots=True)
class Recommendations:
    type: Literal["recommendations"] = "recommendations"
    films: tuple[JSONObject, ...] = ()
    prose: str = ""


@dataclass(frozen=True, slots=True)
class RunFinished:
    type: Literal["run.finished"] = "run.finished"
    stop_reason: str = ""
    turns: int = 0
    tool_calls: int = 0
    input_tokens: int = 0
    output_tokens: int = 0
    cost_usd: float = 0.0
    wall_ms: int = 0


@dataclass(frozen=True, slots=True)
class RunFailed:
    type: Literal["run.failed"] = "run.failed"
    error_code: str = ""
    message: str = ""
    stop_reason: str = ""


type AgentEvent = (
    RunStarted
    | TurnStarted
    | TextDelta
    | ToolCallProposed
    | ToolCallStarted
    | ToolCallFinished
    | PreferenceRecorded
    | GroundingChecked
    | Recommendations
    | RunFinished
    | RunFailed
)

File: web/src/api/events.ts
// The server's event union, mirrored. A parity test compares the member names on both
// sides, so adding an event on the server without handling it here fails CI.

export interface RunStarted {
  type: "run.started"
  run_id: string
  session_id: string
  model: string
  budget: Record<string, number>
}

export interface TurnStarted {
  type: "turn.started"
  turn: number
  tools_offered: string[]
}

export interface TextDelta {
  type: "text.delta"
  text: string
  channel: "thinking" | "answer"
}

export interface ToolCallProposed {
  type: "tool.proposed"
  call_id: string
  name: string
  arguments: Record<string, unknown>
  source: "field" | "content"
}

export interface ToolCallStarted {
  type: "tool.started"
  call_id: string
  name: string
}

export interface ToolCallFinished {
  type: "tool.finished"
  call_id: string
  name: string
  ok: boolean
  summary: string
  meta: Record<string, unknown>
  latency_ms: number
  undo_token: string | null
}

export interface PreferenceRecorded {
  type: "preference.recorded"
  pref_id: number
  label: string
  polarity: string
  hardness: string
  affected_films: number
  undo_token: string
}

export interface GroundingChecked {
  type: "grounding.checked"
  grounded_ratio: number
  nli_available: boolean
  unsupported: string[]
  dropped_film_ids: number[]
}

export interface Recommendations {
  type: "recommendations"
  films: AnsweredFilm[]
  prose: string
}

export interface RunFinished {
  type: "run.finished"
  stop_reason: string
  turns: number
  tool_calls: number
  input_tokens: number
  output_tokens: number
  cost_usd: number
  wall_ms: number
}

export interface RunFailed {
  type: "run.failed"
  error_code: string
  message: string
  stop_reason: string
}

export type AgentEvent =
  | RunStarted
  | TurnStarted
  | TextDelta
  | ToolCallProposed
  | ToolCallStarted
  | ToolCallFinished
  | PreferenceRecorded
  | GroundingChecked
  | Recommendations
  | RunFinished
  | RunFailed

export interface AnsweredFilm {
  film_id: number
  title: string
  year: number | null
  runtime: number | null
  directors: string[]
  why: string
  evidence_refs: string[]
}

const TAGS: ReadonlySet<string> = new Set<AgentEvent["type"]>([
  "run.started",
  "turn.started",
  "text.delta",
  "tool.proposed",
  "tool.started",
  "tool.finished",
  "preference.recorded",
  "grounding.checked",
  "recommendations",
  "run.finished",
  "run.failed",
])

/** One stream frame as a typed event, or null for a tag this build does not know. */
export function asEvent(tag: string, data: Record<string, unknown>): AgentEvent | null {
  if (!TAGS.has(tag) || data.type !== tag) return null
  return data as unknown as AgentEvent
}
